src/udwf.rs

```diff
@@ -0,0 +1,305 @@
+
+            // 2. call function
+            self.evaluator
+                .bind(py)
+                .call_method1("evaluate_all_with_rank", py_args)
+                .map_err(|e| DataFusionError::Execution(format!("{e}")))
+                .map(|v| {
+                    let array_data = ArrayData::from_pyarrow_bound(&v).unwrap();
+                    make_array(array_data)
+                })
+        })
+    }
+
+    fn supports_bounded_execution(&self) -> bool {
+        Python::with_gil(|py| {
+            self.evaluator
+                .bind(py)
+                .call_method0("supports_bounded_execution")
+                .and_then(|v| v.extract())
+                .unwrap_or(false)
+        })
+    }
+
+    fn uses_window_frame(&self) -> bool {
+        Python::with_gil(|py| {
+            self.evaluator
+                .bind(py)
+                .call_method0("uses_window_frame")
+                .and_then(|v| v.extract())
+                .unwrap_or(false)
+        })
+    }
+
+    fn include_rank(&self) -> bool {
+        Python::with_gil(|py| {
+            self.evaluator
+                .bind(py)
+                .call_method0("include_rank")
+                .and_then(|v| v.extract())
+                .unwrap_or(false)
+        })
+    }
+}
+
+pub fn to_rust_partition_evaluator(evaluator: PyObject) -> PartitionEvaluatorFactory {
+    Arc::new(move || -> Result<Box<dyn PartitionEvaluator>> {
+        let evaluator = Python::with_gil(|py| evaluator.clone_ref(py));
+        Ok(Box::new(RustPartitionEvaluator::new(evaluator)))
+    })
+}
+
+/// Represents an WindowUDF
+#[pyclass(name = "WindowUDF", module = "datafusion", subclass)]
+#[derive(Debug, Clone)]
+pub struct PyWindowUDF {
+    pub(crate) function: WindowUDF,
+}
+
+#[pymethods]
+impl PyWindowUDF {
+    #[new]
+    #[pyo3(signature=(name, evaluator, input_types, return_type, volatility))]
+    fn new(
+        name: &str,
+        evaluator: PyObject,
+        input_types: Vec<PyArrowType<DataType>>,
+        return_type: PyArrowType<DataType>,
+        volatility: &str,
+    ) -> PyResult<Self> {
+        let return_type = return_type.0;
+        let input_types = input_types.into_iter().map(|t| t.0).collect();
+
+        let function = WindowUDF::from(MultiColumnWindowUDF::new(
+            name,
+            input_types,
+            return_type,
+            parse_volatility(volatility)?,
+            to_rust_partition_evaluator(evaluator),
+        ));
+        Ok(Self { function })
+    }
+
+    /// creates a new PyExpr with the call of the udf
+    #[pyo3(signature = (*args))]
+    fn __call__(&self, args: Vec<PyExpr>) -> PyResult<PyExpr> {
+        let args = args.iter().map(|e| e.expr.clone()).collect();
+        Ok(self.function.call(args).into())
+    }
+
+    fn __repr__(&self) -> PyResult<String> {
+        Ok(format!("WindowUDF({})", self.function.name()))
+    }
+}
+
+pub struct MultiColumnWindowUDF {
```

Comment:
2 Qs:

1) Does this need to be public?
2) Why motivation behind the name? The `Multi` makes me expect to see a `SingleColumnWindowUDF` counterpart.

Reply:
1. I wanted to make it public in case someone does build extension libraries on top of datafusion-python, but I haven't thought too deeply about which structs need to be public vs private.
2. I called it `MultiColumnWindowUDF` as a contrast to `SimpleWindowUDF` which only takes a single column of input. I'm open to suggestion on a better name.